### DISCUSSION TITLE
ReadCAPTIONImage is very slow

### DISCUSSION THREAD

**Discussion by mrbesen** [ImageMagick/ImageMagick#6675] [Contribute]
### ImageMagick version

f1337d8 (current main)

### Operating system

Linux

### Operating system, version and so on

Linux Mint with Linux 5.15.0-46-generic

### Description

I have created this test program, to measure the time required to construct a `Caption:` image.

https://pastebin.com/V5xCUrkq

On my machine this takes around 1000ms +- 5ms. I think for such a short text and such a simple operation this is way way way to long.

I looked into the source and found the responsible function `ReadCAPTIONImage`  
https://github.com/ImageMagick/ImageMagick/blob/main/coders/caption.c#L107

no surprises here.

I looked further and found the main formatting function `FormatMagickCaption` It often calls `GetTypeMetrics` which is just a wrapper to call `RenderType` with `DrawInfo.render = MagickFalse`
https://github.com/ImageMagick/ImageMagick/blob/main/MagickCore/annotate.c#L929
(in my case with my simple text 1164 times)
And this is the place where most of the time is spent.
Half of it is in the depths of Freetype (i think there is potential to optimize there to get the FontMetrics without actually rendering it).
But what catched my eye are 1130 calls to `GetTypeInfoByFamily` which is using the other half of the processing time.
It looks like the whole block of checking which font to use for the `RenderType` call is executed for each iteration of the main loop of the `FormatMagickCaption` funciton.

My idea to speed this up would be to store the font information, that `RenderType` is using (mainly `const TypeInfo *type_info`) inside the `DrawInfo*` that is given into the `RenderType` function call. So it might be reused by `FormatMagickCaption` on future calls. Maybe even do a separated function to look up the font information at the beginning and store that in `DrawInfo*` to speed up later calls.

I would like to fix that issue, but i am not quite used to the ImageMagick sourcecode and don't know where the functions are used  and a change on them might create issues somewhere else.

Does someone has some tips, where to start and what code to touch or what to NEVER EVER touch?

Here is a screenshot of the callgrind output:
![grafik](https://user-images.githubusercontent.com/8640309/191140189-956d89eb-4071-4f77-92c4-f2688b997d26.png)

I would love to see a 2x speed improvement for the text formatting. I think with some guidance i can get that done.

### Steps to Reproduce

Compile and run the attached program or just use the `caption:` image format.

simple test:
```
time convert -gravity center -size 1024x1024 'caption:testtext abcdefgg kafjlajf jsalkfdj skdjflkj lsakjf lkasjdfaslkdjf lksajfd' abc.png

real	0m0,995s <-- why is it so long?
user	0m1,323s
sys	0m0,052s
```

**Comment by urban-warrior** [comment]
Caption, by default, is an iterative process that tries the text at different font sizes to ensure it optimally fits within the image area.  You can speed up the process by setting appropriate parameters such as the point size, canvas size, etc.  For example,

```
myText->defineValue("caption","start-pointsize","130");
```

This reduces the binary search space and increases performance.   If you can improve the default algorithm, great.  However, in the past, when we tweaked the algorithm, it tended to break certain use cases.  One way to improve performance would be to implement a font glyph cache.  Freetype supports a cache, but we don't currently leverage it in MagickCore/annotate.c/RenderFreetype/1441.
